src/vtcomposite.cpp

```diff
@@ -237,8 +237,8 @@ struct CompositeWorker : Napi::AsyncWorker
         {
             std::string& tile_buffer = *output_buffer_;
             auto buffer = Napi::Buffer<char>::New(
-                Env(),
-                &tile_buffer[0],
+                env,
+                tile_buffer.empty() ? nullptr : &tile_buffer[0],
```

Comment:
Can you add a test that would have failed before this change?

Reply:
In this particular case it's not possible or at least I couldn't see an obvious way to trigger it. On the other hand in `vtshaver` this condition is triggered by an empty list of filters. It was working with node < 14 but segfaults with > 14. 

But there's a more important general principal of "locality" of error handling in code. Accessing internal character buffer of an empty string by taking an address of a first element e.g `&str[0]` is undefined behaviour. In `Execute()` method we have following code block which guarantees that `output_buffer_` is either filled by `vtzero` or `gzip` or an exception is thrown : 

```c++
            std::string& tile_buffer = *output_buffer_;
            if (baton_data_->compress)
            {
                std::string temp;
                builder.serialize(temp);
                tile_buffer = gzip::compress(temp.data(), temp.size());
            }
            else
            {
                builder.serialize(tile_buffer);
            }
```
As long as `vtzero` and `gzip` can never return an empty string all is good. But this is a very weak guarantee if any.

The better way to avoid this particular issue is to replace `std::string` with `std::vector<char>`. This is tricky for the reasons you're aware - explicit use of `std::string` as return type. Another way to address this is to use `c++17` which adds non-const `data()` method to `std::string`. Meanwhile, checking for an empty string is the only reasonable solution. Sorry for a long winded explanation :) And of course we can `const_cast` on `data() but this is even worse I think.

Reply:
👍  sound good, thanks for the explanation.